Address out of order compatibility brands in Filetypebox of heif files

OliverSpeir opened the pull request:
## Changes

1. Add detectType function to `/lib/types/heif.ts` 
  - This improve ability to accurately detect type for HEIF images where the compatibility brands are out of order
    -  Currently `heif.ts` relies on the first brand in the FileTypeBox, but for example some .avif images do not have "avif" as the first type in their FileTypeBox

~2. Sets node version in package.json to `18.4` which shouldn't cause any problems as far as I am aware~

2. add `avis` to the brand map

netroy commented on package.json:
```diff
@@ -27,7 +27,7 @@
   ],
   "engines": {
-    "node": ">=18.18.0"
+    "node": ">=18.4.0"
```
This change should be fine, but I'm curious why this change was needed. 
This package can theoretically work even on nodejs 12 with a few minor changes, but I like to use the very little leverage I have as a package maintainer to push people to upgrade their runtimes, and for that reason I'd like to actually soon bump this to `>=v18.19.1` instead.

OliverSpeir replied:
It's not needed, this would allow Astro to use the package directly instead of patching/vendoring but that is not neccessary, but I will remove it from this PR

netroy commented on lib/types/heif.ts:
```diff
@@ -11,6 +11,24 @@ const brandMap = {
+    }
+  }
+
+  if (brandsDetected.has('avif')) {
+    return 'avif'
+  } else if (brandsDetected.has('heic') || brandsDetected.has('heix') || brandsDetected.has('hevc') || brandsDetected.has('hevx')) {
+    return 'heic'
+  } else if (brandsDetected.has('mif1') || brandsDetected.has('msf1')) {
+    return 'heif'
+  }
```
I see 7 distinct checks here. Do you know if there is an easy way to convert [sample.png](https://github.com/image-size/image-size/blob/main/specs/images/valid/png/sample.png) into test files that we can use to assert these changes?

netroy reviewed (changes requested):
Thanks for this. Can we please add test images to assert these fixes, and to prevent any regressions in the future 🙏🏽

OliverSpeir commented:
what I've done in order to achieve the tests is somewhat hacky, I'll admit. I didn't know how to generate the actual images that would correspond to these conditions, so I manually edited the type boxes of a .avif version of sample.png I created with sharp. 

The way I discovered this condition is from an image in astro's tests. Here's the [original image](https://github.com/withastro/astro/blob/8d5ea2df5d52ad9a311c407533b9f4226480faa8/packages/astro/e2e/fixtures/dev-toolbar/src/light_walrus.avif) and I looked at how this images type box was structured and recreated it. 

All of the test files I added have a very similiar structure, the only difference is to test each condition I changed the second brand. They all are structured like this: 

<img width="240" alt="Screenshot 2024-02-25 at 5 19 42 PM" src="https://github.com/image-size/image-size/assets/115520730/52ac748f-555a-4e4f-8725-37c6fa440484">


@netroy  Hopefully this makes sense and is viable in your opinion I am happy to rework / make any changes you'd like. Thank you

netroy commented:
I've been trying to find a way to create these images properly using a reliable tool. but no luck so far, as no tool lets us define the brands. The manually edited files look okay at a first glance, but the `Image Properties` section in my file browser shows a `Failed to load image information` only for the edited files. No such issues with `light_walrus.avif` though.
So, I'm guessing the manual editing of the brands isn't representative of actual images.
So, I'm conflicted about if I should delete those files, merge the PR, and just hope that this will continue to work as it should, or if I should keep looking for better tools for creating avif/heic/heif.

BTW, I've rebased this on the `main` branch, so that if this gets merged, we don't need to wait for 2.0 to release this to NPM.

OliverSpeir commented:
> or if I should keep looking for better tools for creating avif/heic/heif.

I will try to do the same, I am not too shocked to hear this didn't work I wish I could figure out how the original light_walrus.avif was created, but I'll try to also do some research

OliverSpeir committed:
fix: detect HEIF/HEIC/AVIF type from compatibility brands

Some HEIF-based files use a generic major brand like `mif1` but include
the actual format type (avif, heic, etc.) in their compatibility brands.
This change scans the full ftyp box to correctly identify the format.

Also adds `avis` brand for AVIF sequences.

OliverSpeir commented on lib/types/heif.ts:
```diff
@@ -3,6 +3,7 @@ import { findBox, readUInt32BE, toUTF8String } from './utils'
 
 const brandMap = {
   avif: 'avif',
+  avis: 'avif', // avif-sequence
```
also adds this in here - it was added in https://github.com/withastro/astro/pull/14045 successfully
